src/provider/states/starting.rs

```diff
@@ -56,21 +56,37 @@ impl State<PodState> for Starting {
                     "Starting command: {:?} with arguments {:?}",
                     binary, os_args
                 );
+
+                // Check if environment variables were stated for the container
                 let env_variables = if let Some(vars) = container.env() {
                     debug!(
                         "Got environment vars: {:?} service {}",
                         vars, pod_state.service_name
                     );
-                    vars.iter()
+                    match vars
+                        .iter()
                         .map(|env_var| {
-                            (
-                                String::from(&env_var.name),
-                                String::from(
-                                    &env_var.value.clone().unwrap_or_else(|| String::from("")),
-                                ),
-                            )
+                            let name = String::from(&env_var.name);
+                            let value_result = CreatingConfig::render_config_template(
+                                &template_data,
+                                &env_var.value.clone().unwrap_or_else(|| String::from("")),
+                            );
+                            match value_result {
+                                Ok(value) => Ok((name, value)),
+                                Err(error) => Err(error),
+                            }
```

Comment:
This bit can be changed to this I believe:
```
                            CreatingConfig::render_config_template(
                                &template_data,
                                &env_var.value.clone().unwrap_or_else(String::new),
                            )
                            .map(|value| (name, value))
```

There also must be a better way than cloning the env_var when all you need is a reference anyway.
Doing `env_var.value.as_ref()` works but then you can't return a reference from `unwrap_or_else`.

Reply:
With some help from the Ferrous Rust Experts this is a new suggestion:

```rust
                    let foo = vars
                        .iter()
                        .map(|env_var| {
                            CreatingConfig::render_config_template(
                                &template_data,
                                &env_var.value.as_deref().unwrap_or_default(),
                            )
                            .map(|value| (&env_var.name, value))
                        })
                        .collect();

                    match foo {
                        // All variables were rendered correctly, continue
                        Ok(rendered_value) => rendered_value,
                        // One or more variables failed to render, transition this pod to a failed
                        // state
                        Err(error) => {
                            error!("Failed to render value for env var due to: {:?}", error);
                            return Transition::Complete(Err(anyhow::Error::from(error)));
                        }
                    }
```

Reply:
I've changed the code and added a comment.